From 84f8a3e8e7c53005a678c97eaae4d04a29fd88f4 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 30 Jan 2024 20:01:19 +0100
Subject: [PATCH 0503/1695] drivers: regulator: read current level from the
 device

Always read current voltage level from the device instead of
caching the level in struct regulator. This fixes issues for
when the regulator level value depends on the parent regulator
(supply). It is up the regulator driver to cache or not this
value in their private data if applicable.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d082ed6ddf64a16c381c29cea106b16ad16d0593 drivers: regulator: read current level from the device]
Change-Id: Ib140956cf222f5ba065c7c5858e8a756c07caf4f
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/355960
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/regulator/regulator.c | 39 ++++++++++++++++++------------
 core/include/drivers/regulator.h   |  7 +-----
 2 files changed, 25 insertions(+), 21 deletions(-)

diff --git a/core/drivers/regulator/regulator.c b/core/drivers/regulator/regulator.c
index b7b9f1bf0..92f63f3a0 100644
--- a/core/drivers/regulator/regulator.c
+++ b/core/drivers/regulator/regulator.c
@@ -183,9 +183,27 @@ bool regulator_is_enabled(struct regulator *regulator)
 	return !res && enabled;
 }
 
+int regulator_get_voltage(struct regulator *regulator)
+{
+	TEE_Result res = TEE_SUCCESS;
+	int level_uv = regulator->min_uv;
+
+	if (regulator->ops->get_voltage) {
+		res = regulator->ops->get_voltage(regulator, &level_uv);
+		if (res) {
+			EMSG("%s get_voltage failed with %#"PRIx32,
+			     regulator_name(regulator), res);
+			level_uv = 0;
+		}
+	}
+
+	return level_uv;
+}
+
 TEE_Result regulator_set_voltage(struct regulator *regulator, int level_uv)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
+	int cur_uv = 0;
 
 	assert(regulator);
 	FMSG("%s %duV", regulator_name(regulator), level_uv);
@@ -193,7 +211,8 @@ TEE_Result regulator_set_voltage(struct regulator *regulator, int level_uv)
 	if (level_uv < regulator->min_uv || level_uv > regulator->max_uv)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (level_uv == regulator->cur_uv)
+	cur_uv = regulator_get_voltage(regulator);
+	if (level_uv == cur_uv)
 		return TEE_SUCCESS;
 
 	if (!regulator->ops->set_voltage)
@@ -212,10 +231,10 @@ TEE_Result regulator_set_voltage(struct regulator *regulator, int level_uv)
 	if (regulator->ramp_delay_uv_per_us) {
 		unsigned int d = 0;
 
-		if (regulator->cur_uv > level_uv)
-			d = regulator->cur_uv - level_uv;
+		if (cur_uv > level_uv)
+			d = cur_uv - level_uv;
 		else
-			d = level_uv - regulator->cur_uv;
+			d = level_uv - cur_uv;
 
 		d /= regulator->ramp_delay_uv_per_us;
 
@@ -223,8 +242,6 @@ TEE_Result regulator_set_voltage(struct regulator *regulator, int level_uv)
 		udelay(d);
 	}
 
-	regulator->cur_uv = level_uv;
-
 	return TEE_SUCCESS;
 }
 
@@ -271,15 +288,7 @@ TEE_Result regulator_register(struct regulator *regulator)
 	if (min_uv > max_uv)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	/* Sanitize regulator effective level */
-	if (regulator->ops->get_voltage) {
-		res = regulator->ops->get_voltage(regulator, &uv);
-		if (res)
-			return res;
-	} else {
-		uv = min_uv;
-	}
-	regulator->cur_uv = uv;
+	uv = regulator_get_voltage(regulator);
 
 	if (uv < min_uv || uv > max_uv) {
 		res = regulator_set_voltage(regulator, min_uv);
diff --git a/core/include/drivers/regulator.h b/core/include/drivers/regulator.h
index de162b636..d3ab87049 100644
--- a/core/include/drivers/regulator.h
+++ b/core/include/drivers/regulator.h
@@ -95,7 +95,6 @@ struct regulator_voltages_desc {
  * @max_uv: Max possible voltage level in microvolt (uV)
  * @ramp_delay_uv_per_us: Voltage level change delay in uV/s
  * @enable_ramp_delay_us: Delay after enable, in microseconds (us)
- * @cur_uv: Current voltage level in microvolt (uV)
  * @flags: REGULATOR_* property flags
  * @refcount: Regulator enable request reference counter
  * @lock: Mutex for concurrent access protection
@@ -113,7 +112,6 @@ struct regulator {
 	unsigned int ramp_delay_uv_per_us;
 	unsigned int enable_ramp_delay_us;
 	/* Fields internal to regulator framework */
-	int cur_uv;
 	unsigned int flags;
 	unsigned int refcount;
 	struct mutex lock;	/* Concurrent access protection */
@@ -298,10 +296,7 @@ static inline TEE_Result regulator_set_min_voltage(struct regulator *regulator)
  * regulator_get_voltage() - Get regulator current level in microvolt
  * @regulator: Regulator reference
  */
-static inline int regulator_get_voltage(struct regulator *regulator)
-{
-	return regulator->cur_uv;
-}
+int regulator_get_voltage(struct regulator *regulator);
 
 /*
  * regulator_get_range() - Get regulator min and/or max support levels
-- 
2.47.2

